crates/broker/src/order_picker.rs

```diff
@@ -454,6 +455,46 @@ where
         let pending_balance = self.pending_locked_stake().await?;
         Ok(balance - pending_balance)
     }
+
+    async fn get_pricing_order_capacity(&self) -> Result<Option<u32>> {
+        let max_concurrent_locks = {
+            let config = self.config.lock_all()?;
+            config.market.max_concurrent_locks
+        };
+
+        if let Some(max) = max_concurrent_locks {
+            let committed_orders_count = self.db.get_orders_committed_to_fulfill_count().await?;
+            let available_slots = max.saturating_sub(committed_orders_count);
+            Ok(Some(available_slots))
+        } else {
+            Ok(None)
+        }
+    }
+
+    async fn spawn_pricing_tasks(
+        &self,
+        tasks: &mut JoinSet<Result<Option<U256>, (U256, PriceOrderErr)>>,
+        capacity: u32,
+    ) -> Result<()> {
+        if capacity == 0 {
+            return Ok(());
+        }
+
+        let order_res = self.db.update_orders_for_pricing(capacity).await?;
+
+        for (order_id, order) in order_res {
```

Comment:
It took me a minute to understand mechanically how this PR restricts the number of locks. It appears that is essentially done here since this loop will run 0 times if capacity is zero.

Reply:
yeah, tried to make it simpler, but couldn't think of ways to reduce more. Very open to suggestions, not ecstatic about this current design.

For this, yes, the high-level pattern is checking how many orders can be scheduled, and then spawning that many tasks. In this case though, if capacity is zero the code block isn't hit, but how it restricts further spawning is that the capacity is based on the `JoinSet` of tasks, and unless the pricing fails will count against the capacity, reducing the tasks spawned.